### PR TITLE
Add HeaderQuery and RepoStarred Data

### PR DESCRIPTION
**Pull request by nakaji-dayo**
#198 
I Added `HeaderQuery` in order to customize Accept Header.

For Example, it is needed to get `starred_at` of "List repositories being starred".
https://developer.github.com/v3/activity/starring/#list-repositories-being-starred.

### CODE REVIEW

File: src/GitHub/Data/Request.hs [hunk 1]
```diff
@@ -122,6 +123,7 @@ data Request (k :: Bool) a where
     PagedQuery   :: FromJSON (Vector a) => Paths -> QueryString -> Maybe Count -> Request k (Vector a)
     Command      :: FromJSON a => CommandMethod a -> Paths -> LBS.ByteString -> Request 'True a
     StatusQuery  :: StatusMap a -> Request k () -> Request k a
+    HeaderQuery  :: FromJSON a => Types.RequestHeaders -> Request k a -> Request k a
```
Comment (phadej) [inline src/GitHub/Data/Request.hs]: Clever!

is `FromJSON a` required here, it shouldn't be?

note to myself: refactor the `Request` so we cannot have multiple stacked `HeaderQuery` or `StatusQuery` as it doesn't make sense (or make sure that the code recurse properly).

File: src/GitHub/Request.hs [hunk 2]
```diff
@@ -112,7 +112,9 @@ executeRequestWithMgr :: Manager
                       -> Request k a
                       -> IO (Either Error a)
 executeRequestWithMgr mgr auth req = runExceptT $
-    case req of
+    execute req
+  where
+    execute req' = case req' of
```
Comment (phadej) [inline src/GitHub/Request.hs]: could you give `execute` a type signature?

File: src/GitHub/Request.hs [hunk 3]
```diff
@@ -225,6 +231,9 @@ makeHttpRequest auth r = case r of
                . setBody body
                . setMethod (toMethod m)
                $ req
+    HeaderQuery h req -> do
+        req' <- makeHttpRequest auth req
+        return $ req' { requestHeaders = h <> requestHeaders req'}
```
Comment (phadej) [inline src/GitHub/Request.hs]: space before `}`

File: src/GitHub/Endpoints/Activity/Starring.hs [hunk 4]
```diff
@@ -55,3 +58,13 @@ myStarred auth =
...
+
+
+-- | All the repos starred by the authenticated user.
+myStarredAcceptStar :: Auth -> IO (Either Error (Vector RepoStarred))
+myStarredAcceptStar auth =
+    executeRequest auth $ myStarredAcceptStarR Nothing
+
+-- | All the repos starred by the authenticated user.
+myStarredAcceptStarR :: Maybe Count -> Request 'True (Vector RepoStarred)
+myStarredAcceptStarR mc = HeaderQuery [("Accept", "application/vnd.github.v3.star+json")] $ PagedQuery ["user", "starred"] [] mc
```
Comment (phadej) [inline src/GitHub/Endpoints/Activity/Starring.hs]: you can eta-reduce this: drop `mc` from both sides, change `$` to `.`

File: src/GitHub/Endpoints/Activity/Starring.hs [hunk 5]
```diff
@@ -55,3 +58,13 @@ myStarred auth =
...
 myStarredR :: Maybe Count -> Request 'True (Vector Repo)
 myStarredR = PagedQuery ["user", "starred"] []
+
+
+-- | All the repos starred by the authenticated user.
+myStarredAcceptStar :: Auth -> IO (Either Error (Vector RepoStarred))
+myStarredAcceptStar auth =
+    executeRequest auth $ myStarredAcceptStarR Nothing
+
+-- | All the repos starred by the authenticated user.
```
Comment (phadej) [inline src/GitHub/Endpoints/Activity/Starring.hs]: Please add `see` link. Like in `reposStarredByR`. 

Could you also add to `myStarredR` too, looks like it's missing as well.

**Comment by phadej** [comment]
Please add `myStarredAcceptStarR` to re-exports from `Github` module.

File: src/GitHub/Data/Activities.hs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import Data.Binary              (Binary)
+import Data.Data                (Data, Typeable)
+import GHC.Generics             (Generic)
+import Data.Time                (UTCTime)
+
+import GitHub.Data.Repos (Repo)
+
+data RepoStarred = RepoStarred {
+   repoStarredStarredAt     :: !UTCTime
+  ,repoStarredRepo :: !Repo
```
Comment (phadej) [inline src/GitHub/Data/Activities.hs]: align `::`

File: src/GitHub/Data/Activities.hs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+} deriving (Show, Data, Typeable, Eq, Ord, Generic)
+
+instance NFData RepoStarred where rnf = genericRnf
+instance Binary RepoStarred
+
+-- JSON Instances
+instance FromJSON RepoStarred where
+  parseJSON = withObject "RepoStarred" $ \o ->
```
Comment (phadej) [inline src/GitHub/Data/Activities.hs]: Could this be reformated as
```hs
    parseJSON = withObject "RepoStarred" $ \o -> RepoStarred
        <$> o .: "starred_at"
        <*> o .: "repo"
```

(then there are no non-4-space indentation)

**Comment by phadej** [comment]
Sorry for huge amount of nitpicking, I could do them myself, but I'm not sure when.

Related to this, would you need a Hackage release of this patch? I'd like to hold-on a bit to remember / do other breaking changes as well.

File: src/GitHub/Data/Request.hs (hunk 1 above)
Comment (nakaji-dayo) [inline src/GitHub/Data/Request.hs]: thanks!
you're right. I removed it.

File: src/GitHub/Request.hs (hunk 2 above)
Comment (nakaji-dayo) [inline src/GitHub/Request.hs]: OK, I added it.

File: src/GitHub/Request.hs (hunk 3 above)
Comment (nakaji-dayo) [inline src/GitHub/Request.hs]: fixed it

File: src/GitHub/Endpoints/Activity/Starring.hs (hunk 4 above)
Comment (nakaji-dayo) [inline src/GitHub/Endpoints/Activity/Starring.hs]: oops, my mistake. fixed

File: src/GitHub/Endpoints/Activity/Starring.hs (hunk 5 above)
Comment (nakaji-dayo) [inline src/GitHub/Endpoints/Activity/Starring.hs]: I added to `reposStarredByR` and `myStarredR`

File: src/GitHub/Data/Activities.hs (hunk 6 above)
Comment (nakaji-dayo) [inline src/GitHub/Data/Activities.hs]: I had not know "stylish-haskell". it is awesome. I applied it to all. thanks!

File: src/GitHub/Data/Activities.hs (hunk 7 above)
Comment (nakaji-dayo) [inline src/GitHub/Data/Activities.hs]: reformatted it!

(btw, some files seems to have 2-space indentation. e,g, https://github.com/phadej/github/blob/master/src/GitHub/Data/PullRequests.hs#L267 )

Comment (phadej) [inline src/GitHub/Data/Activities.hs]: Yes, the package isn't formatted consistently. That's unfortunate, at some point I'll reformat it properly.

**Comment by nakaji-dayo** [comment]
No worries. thanks a lot!
I learned a lot from that. And, i'm sorry for my basic mistakes.

> Related to this, would you need a Hackage release of this patch? I'd like to hold-on a bit to remember / do other breaking changes as well.

No, I'm not need Hackage release immediately. 
I can use this from github.

**Comment by nakaji-dayo** [comment]
> Please add myStarredAcceptStarR to re-exports from Github module.
>
> EDIT and Data... to GitHub.Data

I'm sorry, I don't understand.
I have to re-export the GitHub.Data.Activities. Do you need to do something else in the GitHub.Data?

**Comment by phadej** [comment]
Everything was fine. Thanks!

**Pull request merged**